### PR TITLE
BUG: Introduce application properties, differentiate Slicer and main app version

### PR DESCRIPTION
**Pull request by jcfr**
This commit improves how custom application specify there version.
Previously, custom application were overriding variables like Slicer_VERSION_MAJOR
and this was causing build error because the #ifdef" using these variables
were not correct anymore.

With this change custom Slicer application can have their own metadata
(version number, description, ...) in addition of the Slicer ones.

To address this, this commit introduces application properties file and
its companion CMake function "SlicerReadApplicationProperties". This allows
to read the properties of a "Slicer application" before having to process
the corresponding sources. By using this new CMake function in the SlicerApplicationOptions
module included early in the custom application CMakeLists.txt, this
allows to display the application metadata without duplicating much code.

The vtkSlicerVersionConfigure.h was extented to include include these new
macros:
* Slicer_MAIN_PROJECT_VERSION
* Slicer_MAIN_PROJECT_VERSION_FULL
* Slicer_MAIN_PROJECT_BUILDDATE
* Slicer_MAIN_PROJECT_WC_URL
* Slicer_MAIN_PROJECT__WC_REVISION

when building Slicer, these macros are equivalent to Slicer_VERSION,
Slicer_VERSION_FULL, ...

### CODE REVIEW

**Comment by jcfr** [comment]
Failure of CircleCI is expected because variable `Slicer_APPLICATIONS_DIR` is expected to be propagated through the Superbuild. See https://github.com/Slicer/Slicer/pull/997/files#diff-11a4cab360bd5b9583f431dfa45c77e0R18

Error: 
```
CMake Error at CMake/SlicerReadApplicationProperties.cmake:53 (message):
  Slicer_APPLICATIONS_DIR is defined but corresponds to a nonexistent
  directory [Applications]
Call Stack (most recent call first):
  CMake/SlicerApplicationOptions.cmake:31 (SlicerReadApplicationProperties)
  CMakeLists.txt:416 (include)
```

**Comment by jcfr** [comment]
### Next step

* [ ] Rename `Slicer_MAIN_PROJECT_*` variable to `Slicer_MAIN_APP_`. 
  * Indeed, Slicer can support having multiple applications sources in the `Application` folder, and one of them can be set as the "main" one. Setting an application as the main one defines which metadata are used to create the installer.

* [ ] Move about dialog to `Base/QtApp` library
* [ ] Move Main windows to `Base/QtApp` library

**Comment by jcfr** [comment]
> Rename Slicer_MAIN_PROJECT_* variable to Slicer_MAIN_APP_.

This is tracked by https://issues.slicer.org/view.php?id=4588 and will be implemented later

**Comment by jcfr** [comment]
After doing a clean build, I address one last issue. This is now ready for integration.

**Comment by jcfr** [comment]
I will plan to integrate around noon.

File: CMake/vtkSlicerVersionConfigure.h.in [hunk 1]
```diff
@@ -33,17 +38,39 @@
...
+#define Slicer_MAIN_PROJECT_BUILDDATE "@Slicer_MAIN_PROJECT_BUILDDATE@"
+
+/// \def Slicer_MAIN_PROJECT_WC_URL
+/// \brief Source control management URL identifying main project sources.
+#define Slicer_MAIN_PROJECT_WC_URL "@Slicer_MAIN_PROJECT_WC_URL@"
+
+/// \def Slicer_MAIN_PROJECT_WC_REVISION
+/// \brief Revision identifying main project sources.
+/// \sa Slicer_MAIN_PROJECT_WC_URL
+#define Slicer_MAIN_PROJECT__WC_REVISION "@Slicer_MAIN_PROJECT_WC_REVISION@"
```
Comment (cpinter) [inline CMake/vtkSlicerVersionConfigure.h.in]: Are two underscores intentional here?

Comment (jcfr) [reply CMake/vtkSlicerVersionConfigure.h.in]: Nop. Good catch

Comment (jcfr) [reply CMake/vtkSlicerVersionConfigure.h.in]: And thanks for the review :smile:

**Comment by jcfr** [comment]
Closing. Changes integrated in [r27316](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=27316) to [r27324](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=27324)

**Pull request closed without merging**